evidence/store.go

```diff
@@ -0,0 +1,189 @@
+	return &EvidenceStore{
+		db: db,
+	}
+}
+
+// PriorityEvidence returns the evidence from the outqueue, sorted by highest priority.
+func (store *EvidenceStore) PriorityEvidence() (evidence []types.Evidence) {
+	// reverse the order so highest priority is first
+	l := store.ListEvidence(baseKeyOutqueue)
+	l2 := make([]types.Evidence, len(l))
+	for i := range l {
+		l2[i] = l[len(l)-1-i]
+	}
+	return l2
+}
+
+// PendingEvidence returns all known uncommitted evidence.
+func (store *EvidenceStore) PendingEvidence() (evidence []types.Evidence) {
+	return store.ListEvidence(baseKeyPending)
+}
+
+// ListEvidence lists the evidence for the given prefix key.
+// It is wrapped by PriorityEvidence and PendingEvidence for convenience.
+func (store *EvidenceStore) ListEvidence(prefixKey string) (evidence []types.Evidence) {
+	iter := store.db.IteratorPrefix([]byte(prefixKey))
+	for iter.Next() {
+		val := iter.Value()
+
+		var ei EvidenceInfo
+		wire.ReadBinaryBytes(val, &ei)
+		evidence = append(evidence, ei.Evidence)
+	}
+	return evidence
+}
+
+// GetEvidence fetches the evidence with the given height and hash.
+func (store *EvidenceStore) GetEvidence(height int, hash []byte) *EvidenceInfo {
+	key := keyLookupFromHeightAndHash(height, hash)
+	val := store.db.Get(key)
+
+	if len(val) == 0 {
+		return nil
+	}
+	var ei EvidenceInfo
+	wire.ReadBinaryBytes(val, &ei)
+	return &ei
+}
+
+// AddNewEvidence adds the given evidence to the database.
+// It returns false if the evidence is already stored.
+func (store *EvidenceStore) AddNewEvidence(evidence types.Evidence, priority int) bool {
+	// check if we already have seen it
+	ei_ := store.GetEvidence(evidence.Height(), evidence.Hash())
+	if ei_ != nil && ei_.Evidence != nil {
+		return false
+	}
+
+	ei := EvidenceInfo{
+		Committed: false,
+		Priority:  priority,
+		Evidence:  evidence,
+	}
+	eiBytes := wire.BinaryBytes(ei)
+
+	// add it to the store
+	key := keyOutqueue(evidence, priority)
+	store.db.Set(key, eiBytes)
+
+	key = keyPending(evidence)
+	store.db.Set(key, eiBytes)
+
+	key = keyLookup(evidence)
+	store.db.SetSync(key, eiBytes)
+
+	return true
+}
+
+// MarkEvidenceAsBroadcasted removes evidence from Outqueue.
+func (store *EvidenceStore) MarkEvidenceAsBroadcasted(evidence types.Evidence) {
+	ei := store.getEvidenceInfo(evidence)
+	key := keyOutqueue(evidence, ei.Priority)
+	store.db.Delete(key)
+}
+
+// MarkEvidenceAsPending removes evidence from pending and outqueue and sets the state to committed.
+func (store *EvidenceStore) MarkEvidenceAsCommitted(evidence types.Evidence) {
+	// if its committed, its been broadcast
+	store.MarkEvidenceAsBroadcasted(evidence)
+
+	pendingKey := keyPending(evidence)
+	store.db.Delete(pendingKey)
+
+	ei := store.getEvidenceInfo(evidence)
+	ei.Committed = true
+
+	// TODO: we should use the state db and db.Sync in state.Save instead.
+	// Else, if we call this before state.Save, we may never mark committed evidence as committed.
```

Comment:
We should just mark it as committed before state.Save.  It'll get committed again if we restart.

Reply:
There's a case during replay if the app is multiple blocks behind that we don't construct historical state and call state.ApplyBlock, we use sm.ExecCommitBlock instead, so eg. MarkEvidenceAsCommitted would never get called. 

But now with historical validators etc we should be able to construct the historical state to use ApplyBlock after all.